[68560] Fix parent dates not updated when updating dates on child with automatically generated subject

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/68560

# What are you trying to accomplish?

Fix the following bug: if a work package has a type with automatically generated subject, updating its dates won't reschedule its ancestors and successors.

# What approach did you choose and why?

The problem was that the `WorkPackages::UpdateService` updates the related work packages in two phases:

- First it updates the ancestors, including itself. This updates work, remaining work, percent complete, switching to automatic mode, and ignore non-working days attributes.

- Second, if there are changes to dates, it updates the work packages to reschedule: the ancestors and successors. This updates the start date, finish date, and duration attributes.

This means that if something is changed in the first phase, the initiating work package will be saved and there will be no changes left to detect that a rescheduling of ancestors is needed too.

When having work packages with automatically generated subjects, that's what happens: the subject is always updated to a placeholder text ("Automatically generated through type <type name>"), leading to the `UpdateAncestorsService` always detecting a change and so always saving it. This inhibits the rescheduling of ancestors because there are no changes in the initiator work package.

~The fix is to have a stricter change detection in the `UpdateAncestorsService` to avoid saving if nothing was changed by the service itself.~

~⚠️ It's probable that modifying some attributes along with dates could lead to the same problem. For instance modifying both work (estimated hours) and dates at the same time using the REST API will probably not reschedule ancestors and successors. As long it's not attributes touched by the `UpdateAncestorsService`, it should work properly.~

Fix was not working properly because automatically generated subjects rely on `UpdateAncestorsService` to be saved.

New fix is to consider both changes from `UpdateService` and `UpdateAncestorsService` when calling `SetScheduleService`.

BTW I'm not found of this two phase approach, but the rescheduling needs some information to be saved in database so that the rescheduling can work properly. Changing that would need a bigger refactoring.

That seems correlated to [Bug 61758: Duplicated journal entries when updating attributes of related work package](https://community.openproject.org/projects/stream-planning-and-reporting/work_packages/61758) and it's possible that fixing that bug will fix the 2-phase issue.

### automatically generated subject integration test

An integration test was added to be able to detect side-effects where it's not saved. This was missing.

### Changes in the table helpers

As the subject is automatically generated in the related integration test, the subject can't be used to find the work packages having been created. So I added the ability to use only the identifier in the assertion: the ids of the work packages created during the setup phase are saved along with the identifier having been used, so that when later retrieving the work package to checking their attributes, they can be referenced by the identifier used when creating them instead of the subject (which has been automatically updated as soon as it was saved).

This was necessary to write a test in `spec/services/work_packages/update_service_integration_spec.rb` using the table helpers.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
Add ability to use identifiers in table helpers

This allows to use identifiers in table helpers without checking the
subjects. That's useful for work packages with types having
automatically generated subjects.

cbliard committed:
[68560] Fix change detection in update ancestors service

Fixes https://community.openproject.org/wp/68560

The problem was that the `WorkPacakges::UpdateService` updates the
related work packages in two phases:

- First it updates the ancestors, including itself. This updates work,
  remaining work, percent complete, switching to automatic mode, and
  ignore non-working days attributes.

- Second, if there are changes to dates, it updates the work packages to
  reschedule: the ancestors and successors. This updates the start date,
  finish date, and duration attributes.

This means that if something is changed in the first phase, the
initiating work package will be saved and there will be no changes left
to detect that a rescheduling of ancestors is needed too.

When having work packages with automatically generated subjects, that's
what happens: the subject is always updated to a placeholder text
("Automatically generated through type <type name>"), leading to the
`UpdateAncestorsService` always detecting a change and so always saving
it. This inhibits the rescheduling of ancestors.

The fix is to have a stricter change detection in the
`UpdateAncestorsService` to avoid saving if nothing was changed by the
service itself.

It feels brittle to me, and it's probable that modifying some attributes
along with dates could lead to the same problem. For instance modifying
work (estimated hours) and dates at the same time using the REST API.

Kharonus commented on spec/services/work_packages/update_service_integration_spec.rb:
```diff
@@ -1225,6 +1225,39 @@
+
+      it "updates the dates of the parent" do
+        expect(subject).to be_success
+        # expect(subject.all_results.pluck(:id)).to contain_exactly(work_package.id, parent.id)
```
🟡 left a commented line here by accident?

cbliard replied:
Good catch

Kharonus commented on spec/services/work_packages/update_service_integration_spec.rb:
```diff
@@ -1225,6 +1225,39 @@
+        expect(subject).to be_success
+        # expect(subject.all_results.pluck(:id)).to contain_exactly(work_package.id, parent.id)
+
+        expect_work_packages_after_reload([parent, child, work_package], <<~TABLE)
```
🟢 nice spec setup!

cbliard replied:
Thanks 🙌
The table helpers took me a while to write but they make writing new tests fast and relatively easy to read.

Kharonus reviewed (approved):
fixes the current bug nicely! The related problems you have already described. LGTM after fixing currently failing unit tests

cbliard committed:
[68560] Reschedule related work packages even if save occurs during UpdateAncestors

https://community.openproject.org/wp/68560

Previous fix of the same issue was incorrect, because
`UpdateAncestorsService` was no longer saving the work package if the
changes were done outside of the service. Sadly, automatically generated
subjects are updated outside of the service and rely on
`UpdateAncestorsService` to be saved. That's a code smell which should
be fixed in a future refactoring.

An integration test was added to ensure the automatically generated
subject is saved correctly, the fix was reverted, and another fix was
found.

Now the fix is to consider the changes from both the update service and
the `UpdateAncestorsService` when calling the `SetScheduleService`.

cbliard committed:
Fix rubocop warning

cbliard committed:
Fix warnings about redefinition of let(:parent)

cbliard committed:
Add another unit test for automatically generated subjects

Kharonus commented on app/services/work_packages/shared/update_ancestors.rb:
```diff
@@ -31,8 +31,8 @@
     module UpdateAncestors
-      def update_ancestors(changed_work_packages)
-        changes = changed_work_packages
+      def update_ancestors(changed_work_packages, changes = nil)
```
❓ seeing this, makes me wonder. Are there any occurrences where it is important to not pass the changes into the method, but fetch them in the next line?

cbliard replied:
actually no, that's the only place where it's called.
Nice catch. I'll rework it.

cbliard replied:
That was a fun refactoring :)

Kharonus commented on app/services/work_packages/update_service.rb:
```diff
@@ -56,19 +60,24 @@ def after_perform(service_call)
 
-    update_related(service_call.result).each do |related_service_call|
+    # update saved changes as they might have changed due to the ancestors updates
+    changed_attributes += work_package.changed_attribute_keys_before_last_save
```
🟢 nice idea

Kharonus reviewed (approved):
Just looking from code perspective. LGTM.

no local test run :D Just because we had this talk last time. If you need a dev testing this locally, please tell me.

I have one question, that might trigger a rework, but IMHO this is not a blocker

cbliard committed:
refactor: change parameters shape of `update_ancestors` methods

- `#update_ancestors` was never called with multiple work packages at
  once. So now it accepts only one work package.
- `#update_ancestors_all_attributes` was called only twice and accepts
  multiple work packages. To differentiate it from `#update_ancestors`,
  prefix it with `multi_`.
- remove the need to have both `#multi_update_ancestors` and
  `#multi_update_ancestors_all_attributes` as it's always called with
  all attributes, so keep only `#multi_update_ancestors` and make
  `#update_ancestors` use all attributes if no changed attributes are
  provided.
- then clean up places where it was used.
